pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaMPI.hpp

```diff
@@ -0,0 +1,2039 @@
+
+    /**
+     * @brief Normalize the index ordering to match PennyLane.
+     *
+     * @tparam IndexType Integer value type.
+     * @param indices Given indices to transform.
+     */
+    template <typename IndexType>
+    inline auto NormalizeIndices(std::vector<IndexType> indices)
+        -> std::vector<IndexType> {
+        std::vector<IndexType> t_indices(std::move(indices));
+        std::transform(t_indices.begin(), t_indices.end(), t_indices.begin(),
+                       [&](IndexType i) -> IndexType {
+                           return BaseType::getNumQubits() - 1 - i;
+                       });
+        return t_indices;
+    }
+
+    /**
+     * @brief Get expectation value for a sum of Pauli words.
+     *
+     * @param pauli_words Vector of Pauli-words to evaluate expectation value.
+     * @param tgts Coupled qubit index to apply each Pauli term.
+     * @param coeffs Numpy array buffer of size |pauli_words|
+     * @return auto Expectation value.
+     */
+    auto expvalOnPauliBasis(const std::vector<std::string> &pauli_words,
+                            const std::vector<std::vector<std::size_t>> &tgts,
+                            std::vector<double> &local_expect) {
+        uint32_t nIndexBits = static_cast<uint32_t>(this->getNumLocalQubits());
+        cudaDataType_t data_type;
+
+        if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                      std::is_same_v<CFP_t, double2>) {
+            data_type = CUDA_C_64F;
+        } else {
+            data_type = CUDA_C_32F;
+        }
+
+        // Note: due to API design, cuStateVec assumes this is always a double.
+        // Push NVIDIA to move this to behind API for future releases, and
+        // support 32/64 bits.
+
+        std::vector<std::vector<custatevecPauli_t>> pauliOps;
+
+        std::vector<custatevecPauli_t *> pauliOps_ptr;
+
+        for (auto &p_word : pauli_words) {
+            pauliOps.push_back(cuUtil::pauliStringToEnum(p_word));
+            pauliOps_ptr.push_back((*pauliOps.rbegin()).data());
+        }
+
+        std::vector<std::vector<int32_t>> basisBits;
+        std::vector<int32_t *> basisBits_ptr;
+        std::vector<uint32_t> n_basisBits;
+
+        for (auto &wires : tgts) {
+            std::vector<int32_t> wiresInt(wires.size());
+            std::transform(wires.begin(), wires.end(), wiresInt.begin(),
+                           [&](std::size_t x) { return static_cast<int>(x); });
+            basisBits.push_back(wiresInt);
+            basisBits_ptr.push_back((*basisBits.rbegin()).data());
+            n_basisBits.push_back(wiresInt.size());
+        }
+
+        // compute expectation
+        PL_CUSTATEVEC_IS_SUCCESS(custatevecComputeExpectationsOnPauliBasis(
+            /* custatevecHandle_t */ handle_.get(),
+            /* void* */ BaseType::getData(),
+            /* cudaDataType_t */ data_type,
+            /* const uint32_t */ nIndexBits,
+            /* double* */ local_expect.data(),
```

Comment:
Looks like this is still expecting double-prec here. Ignore all earlier comments if this is the limitation.

Reply:
yes, we still need double here for the expect result but we can use fp32 for the sv!